File: app/static/js/readPixelsFromJson.js
const fs = require('fs');
const path = require('path');
const sharp = require('sharp');

// const profileName = process.argv[2]; // Erwartet den Profilnamen als CLI-Argument
// console.log('ProfilName readPixels', profileName);
// console.log('fs Path:', path);
const profileName = "elbenwald";

if (!profileName) {
    console.error('Kein Profilname angegeben! Das Skript benötigt einen Profilnamen.');
    process.exit(1);
}
function wait(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

console.log("analyzeColors() gestartet.");
const jsonFile = path.join(__dirname, `${profileName}_images.json`);

console.log("Vor dem Laden der Datei");
let files;
try {
    const fileData = fs.readFileSync(jsonFile, 'utf8');
    files = JSON.parse(fileData); // Bildliste aus JSON-Datei
    console.log('Dateien aus JSON geladen:', files);
} catch (error) {
    console.error("Fehler: JSON-Datei konnte nicht geladen werden.", error);
    process.exit(1);
}

const colorData = {};

async function processImage(filePath) {
    try {
        const imageBuffer = fs.readFileSync(filePath);
        console.log('Bild geladen, Größe:', imageBuffer.length);
        console.log('Imagebuffer::', imageBuffer);

        if (imageBuffer.length === 0) {
            console.error('Das Bild ist leer oder konnte nicht geladen werden.');
            return;
        } else {
            console.log('Bild erfolgreich geladen.');
        }

        // Verzeichnis für bearbeitete Bilder sicherstellen
        const processedImagesDir = path.join(__dirname, 'processed_images');
        if (!fs.existsSync(processedImagesDir)) {
            fs.mkdirSync(processedImagesDir);
        }

        // Verarbeite und speichere das Bild
        const outputFilePath = path.join(processedImagesDir, path.basename(filePath));
        
        // Verarbeite das Bild mit sharp und speichere es, während der Code auf das Ergebnis wartet
        await sharp(imageBuffer)
            .resize(200, 200)
            .toFile(outputFilePath);
        
        console.log('Bild gespeichert:', outputFilePath);

        await wait(10000); // Wartet 10 Sekunden (10000 Millisekunden)
    
        // Optionale Analyse des bearbeiteten Bildes, falls gewünscht
        const { data, info } = await sharp(outputFilePath)
            .raw()
            .toBuffer();

        console.log("Buffer verarbeitet:", outputFilePath);
        console.log("Data:", data);
        console.log("Bildinfo:", info);  // Gibt Aufschluss über die Bilddetails
        console.log("Data length:", data.length);

        if (data.length === 0) {
            console.error("Data array is empty!");
        }

        const colors = {};
        for (let i = 0; i < data.length; i += 4) {
            const r = data[i];
            const g = data[i + 1];
            const b = data[i + 2];
            const hex = `#${((1 << 24) + (r << 16) + (g << 8) + b).toString(16).slice(1)}`;
            colors[hex] = (colors[hex] || 0) + 1;
        }

        colorData[filePath] = mergeSimilarColors(colors);
        console.log('Aktuelle Farbanalyse für Datei:', filePath, colorData[filePath]);
    } catch (err) {
        console.error('Fehler beim Verarbeiten der Datei:', filePath, err);
    }
}


async function processImages(files) {
    const promises = files.map(file => {
        const shortenedPath = path.resolve(__dirname, '..', '..');
        const filePath = path.join(shortenedPath, file);
        console.log('Verarbeite Datei:', filePath);
        return processImage(filePath);
    });

    await Promise.all(promises);
    saveColorDataToFile(colorData, profileName);
    console.log('!!!!!!!color Daten:', colorData);
    console.log("Pixel-Farben erfolgreich gespeichert.");
}

processImages(files).catch(console.error);

function mergeSimilarColors(colors) {
    const mergedColors = {};
    const threshold = 30;

    for (const color in colors) {
        const [r1, g1, b1] = hexToRgb(color);
        let found = false;

        for (const mergedColor in mergedColors) {
            const [r2, g2, b2] = hexToRgb(mergedColor);
            if (
                Math.abs(r1 - r2) < threshold &&
                Math.abs(g1 - g2) < threshold &&
                Math.abs(b1 - b2) < threshold
            ) {
                mergedColors[mergedColor] += colors[color];
                found = true;
                break;
            }
        }

        if (!found) {
            mergedColors[color] = colors[color];
        }
    }

    return mergedColors;
}

function hexToRgb(hex) {
    const bigint = parseInt(hex.slice(1), 16);
    const r = (bigint >> 16) & 255;
    const g = (bigint >> 8) & 255;
    const b = bigint & 255;
    return [r, g, b];
}

function saveColorDataToFile(colorData, profileName) {
    const filePath = path.join(__dirname, `${profileName}_color_data.json`);
    try {
        fs.writeFileSync(filePath, JSON.stringify(colorData, null, 2), 'utf8');
        console.log(`Daten erfolgreich in ${filePath} gespeichert.`);
    } catch (err) {
        console.error('Fehler beim Speichern der JSON-Datei:', err);
    }
}

File: app/readPixelsFromJson.py
import os
import json
from PIL import Image
from collections import defaultdict

# Name des Profilordners für die Eingabedateien
if len(sys.argv) < 2:
    print("Profilname fehlt!")
    sys.exit(1)

profile_name = sys.argv[1]


# Verzeichnis des aktuellen Skripts ermitteln
current_dir = os.path.dirname(os.path.abspath(__file__))

# Verzeichnis der Bilder basierend auf dem Speicherort des Skripts festlegen
images_dir = os.path.join(current_dir, 'static', 'images', 'elbenwald_Pics')

# JSON-Dateipfad für die Liste der Bilddateien
json_file_path = os.path.join(current_dir, f"{profile_name}_images.json")

# Ausgabe-Pfad für die Farbanalyse in static/js
output_dir = os.path.join(current_dir, 'static', 'js')
if not os.path.exists(output_dir):
    os.makedirs(output_dir)

# Ergebnis-Dictionary für Farbanalysen
color_data = {}

def load_image_files(json_file_path):
    """Lädt eine Liste von Bilddateien aus einer JSON-Datei."""
    try:
        with open(json_file_path, 'r') as file:
            files = json.load(file)
        print("Dateien aus JSON geladen:", files)
        return files
    except (FileNotFoundError, json.JSONDecodeError) as e:
        print("Fehler beim Laden der JSON-Datei:", e)
        return []

def process_image(file_path):
    """Verarbeitet ein Bild und zählt die Farben der Pixel."""
    try:
        with Image.open(file_path) as img:
            img = img.resize((200, 200))  # Optionales Resizing des Bildes
            img = img.convert("RGB")  # Bild in RGB umwandeln

            pixels = img.getdata()
            color_counts = defaultdict(int)

            for pixel in pixels:
                hex_color = "#{:02x}{:02x}{:02x}".format(pixel[0], pixel[1], pixel[2])
                color_counts[hex_color] += 1

            color_data[file_path] = merge_similar_colors(color_counts)
            print(f"Farbanalyse für {file_path}: {color_data[file_path]}")

    except Exception as e:
        print(f"Fehler beim Verarbeiten der Datei {file_path}: {e}")

def merge_similar_colors(colors, threshold=30):
    """Merges colors that are similar within a certain threshold."""
    merged_colors = {}

    for color, count in colors.items():
        r1, g1, b1 = hex_to_rgb(color)
        found = False

        for merged_color in merged_colors:
            r2, g2, b2 = hex_to_rgb(merged_color)
            if (abs(r1 - r2) < threshold) and (abs(g1 - g2) < threshold) and (abs(b1 - b2) < threshold):
                merged_colors[merged_color] += count
                found = True
                break

        if not found:
            merged_colors[color] = count

    return merged_colors

def hex_to_rgb(hex_color):
    """Konvertiert eine hexadezimale Farbe in RGB."""
    hex_color = hex_color.lstrip('#')
    return int(hex_color[0:2], 16), int(hex_color[2:4], 16), int(hex_color[4:6], 16)

def save_color_data_to_file(color_data, profile_name):
    """Speichert die Farbanalyse in einer JSON-Datei im static/js-Verzeichnis."""
    output_path = os.path.join(output_dir, f"{profile_name}_color_data.json")
    try:
        with open(output_path, 'w', encoding='utf-8') as file:
            json.dump(color_data, file, indent=2, ensure_ascii=False)
        print(f"Daten erfolgreich in {output_path} gespeichert.")
    except IOError as e:
        print(f"Fehler beim Speichern der Datei {output_path}: {e}")

def process_images(files):
    """Verarbeitet alle Bilder in der angegebenen Liste."""
    for file_path in files:
        full_path = os.path.join(images_dir, file_path)
        print(f"Verarbeite Datei: {full_path}")
        process_image(full_path)
    save_color_data_to_file(color_data, profile_name)

if __name__ == "__main__":
    files = load_image_files(json_file_path)
    if files:
        process_images(files)